cpp/src/centrality/betweenness_centrality_impl.cuh

```diff
@@ -538,6 +542,720 @@ void accumulate_edge_results(
+      vertex_t dist             = distances[v_offset];
+
+      if (dist >= 0 && dist <= global_max_distance) {
+        cuda::atomic_ref<size_t, cuda::thread_scope_device> counter(distance_counts[dist]);
+        counter.fetch_add(size_t{1}, cuda::std::memory_order_relaxed);
+      }
+    });
+
+  // Copy counts to host and allocate buckets
+  std::vector<size_t> host_distance_counts(global_max_distance + 1);
+  raft::update_host(host_distance_counts.data(),
+                    distance_counts.data(),
+                    global_max_distance + 1,
+                    handle.get_stream());
+  handle.sync_stream();
+
+  // Create buckets for each distance level
+  std::vector<rmm::device_uvector<vertex_t>> distance_buckets_vertices;
+  std::vector<rmm::device_uvector<origin_t>> distance_buckets_sources;
+
+  // Reserve space for all distance levels
+  distance_buckets_vertices.reserve(global_max_distance + 1);
+  distance_buckets_sources.reserve(global_max_distance + 1);
+
+  // Allocate exact-sized buckets
+  for (vertex_t d = 0; d <= global_max_distance; ++d) {
+    if (host_distance_counts[d] > 0) {
+      distance_buckets_vertices.emplace_back(host_distance_counts[d], handle.get_stream());
+      distance_buckets_sources.emplace_back(host_distance_counts[d], handle.get_stream());
+    } else {
+      distance_buckets_vertices.emplace_back(0, handle.get_stream());
+      distance_buckets_sources.emplace_back(0, handle.get_stream());
+    }
+  }
+
+  // Reset counts for use as offsets
+  thrust::fill(
+    handle.get_thrust_policy(), distance_counts.begin(), distance_counts.end(), size_t{0});
+
+  auto v_first = graph_view.local_vertex_partition_range_first();
+
+  // Create arrays of raw pointers for device access
+  std::vector<vertex_t*> host_bucket_vertices_ptrs(global_max_distance + 1);
+  std::vector<origin_t*> host_bucket_sources_ptrs(global_max_distance + 1);
+
+  for (vertex_t d = 0; d <= global_max_distance; ++d) {
+    host_bucket_vertices_ptrs[d] = distance_buckets_vertices[d].data();
+    host_bucket_sources_ptrs[d]  = distance_buckets_sources[d].data();
+  }
+
+  // Copy pointer arrays to device
+  rmm::device_uvector<vertex_t*> device_bucket_vertices_ptrs(global_max_distance + 1,
+                                                             handle.get_stream());
+  rmm::device_uvector<origin_t*> device_bucket_sources_ptrs(global_max_distance + 1,
+                                                            handle.get_stream());
+
+  raft::update_device(device_bucket_vertices_ptrs.data(),
+                      host_bucket_vertices_ptrs.data(),
+                      global_max_distance + 1,
+                      handle.get_stream());
+  raft::update_device(device_bucket_sources_ptrs.data(),
+                      host_bucket_sources_ptrs.data(),
+                      global_max_distance + 1,
+                      handle.get_stream());
+
+  // Ensure pointer arrays are copied before kernel launch
+  handle.sync_stream();
```

Comment:
No need here as this host to copy operation and the following device kernel launch both run on the same stream.

Reply:
In general, you need to call handle.sync_stream() when you copy data to host and manipulate the data in host (as the host code is not stream ordered) but you don't need this when you copy data to device and launch a device kernel (unless they execute on different streams)